esmf_regrid/tests/unit/schemes/test_ESMFNearestRegridder.py

```diff
@@ -0,0 +1,318 @@
+    curv_src_rad = _curvilinear_cube(
+        n_lons_src, n_lats_src, lon_rad_bounds, lat_rad_bounds
+    )
+    rad_coords(curv_src_rad)
+    curv_tgt = _curvilinear_cube(n_lons_tgt, n_lats_tgt, lon_bounds, lat_bounds)
+    curv_tgt_rad = _curvilinear_cube(
+        n_lons_tgt, n_lats_tgt, lon_rad_bounds, lat_rad_bounds
+    )
+    rad_coords(curv_tgt_rad)
+
+    grid_to_grid = ESMFNearestRegridder(grid_src, grid_tgt)
+    grid_rad_to_grid = ESMFNearestRegridder(grid_src_rad, grid_tgt)
+    grid_rad_to_curv = ESMFNearestRegridder(grid_src_rad, curv_tgt)
+    curv_to_grid_rad = ESMFNearestRegridder(curv_src, grid_tgt_rad)
+    curv_rad_to_grid = ESMFNearestRegridder(curv_src_rad, grid_tgt)
+    curv_to_curv_rad = ESMFNearestRegridder(curv_src, curv_tgt_rad)
+
+    def extract_weights(regridder):
+        return regridder.regridder.weight_matrix.todense()
+
+    for regridder in [
+        grid_rad_to_grid,
+        grid_rad_to_curv,
+        curv_to_grid_rad,
+        curv_rad_to_grid,
+        curv_to_curv_rad,
+    ]:
+        assert np.allclose(extract_weights(grid_to_grid), extract_weights(regridder))
+
+
+def test_masks():
+    """
+    Test initialisation of :func:`esmf_regrid.schemes.ESMFNearestRegridder`.
+
+    Checks that the `use_src_mask` and `use_tgt_mask` keywords work properly.
+    """
+    src = _curvilinear_cube(7, 6, [-180, 180], [-90, 90])
+    tgt = _curvilinear_cube(6, 7, [-180, 180], [-90, 90])
+
+    # Make src and tgt discontiguous at (0, 0)
+    src_mask = np.zeros([6, 7], dtype=bool)
+    src_mask[0, 0] = True
+    src.data = np.ma.array(src.data, mask=src_mask)
+    src_discontiguous = src.copy()
+    src_discontiguous.coord("latitude").bounds[0, 0] = 0
+    src_discontiguous.coord("longitude").bounds[0, 0] = 0
+
+    tgt_mask = np.zeros([7, 6], dtype=bool)
+    tgt_mask[0, 0] = True
+    tgt.data = np.ma.array(tgt.data, mask=tgt_mask)
+    tgt_discontiguous = tgt.copy()
+    tgt_discontiguous.coord("latitude").bounds[0, 0] = 0
+    tgt_discontiguous.coord("longitude").bounds[0, 0] = 0
+
+    rg_src_masked = ESMFNearestRegridder(src_discontiguous, tgt, use_src_mask=True)
+    rg_tgt_masked = ESMFNearestRegridder(src, tgt_discontiguous, use_tgt_mask=True)
+    rg_unmasked = ESMFNearestRegridder(src, tgt)
+
+    weights_src_masked = rg_src_masked.regridder.weight_matrix
+    weights_tgt_masked = rg_tgt_masked.regridder.weight_matrix
+    weights_unmasked = rg_unmasked.regridder.weight_matrix
+
+    # Check there are no weights associated with the masked point.
+    assert weights_src_masked[:, 0].nnz == 0
+    assert weights_tgt_masked[0].nnz == 0
+
+    # Check other weights are correct. Note that unique to NEAREST_DTOS, masking a source
+    # point causes the next nearest source point to gain weights in the weight matrix.
+    # because of this, we ignore the row associated with that target point and check
+    # the rest of the weights matrix.
```

Comment:
I don't follow this comment. This test is actually checking _more_ of the weights matrix than its equivalent in `test_ESMFBilinearRegridder`, and the same as `test_ESMFAreaWeightedRegridder` which does not have an equivalent comment.

Have I missed something? Or is there a better phrasing for this comment that the uninitiated might better understand?

Reply:
This is not the same as the `test_ESMFAreaWeightedRegridder` test. compare the line:
```python
assert np.allclose(weights_src_masked[1:].todense(), weights_unmasked[1:].todense())
```
with

```python
assert np.allclose(
        weights_src_masked[:, 1:].todense(), weights_unmasked[:, 1:].todense()
    )
```

For `test_ESMFNearestRegridder `, everything associated with the first source point is ignored, for `test_ESMFAreaWeightedRegridder ` everything assoicated with either the masked source *or* the masked target point is ignored.

Reply:
Thanks for straightening me out